xdsl/ir.py

```diff
@@ -942,6 +947,16 @@ def is_structurally_equivalent(
 
         return True
 
+    def emit_error(
+        self, message: str, exception_type: type[Exception] = VerifyException
+    ) -> NoReturn:
+        """Emit an error with the given message."""
+        from xdsl.utils.diagnostic import Diagnostic
```

Comment:
Why the inline import? Circular dependency?

Reply:
Maybe add a TODO. I don't like circular dependencies too much but I guess getting rid of it for real would be quite some restructuring, so let's leave that for later, ig.

Reply:
Why though?
I don't know how I feel about putting a TODO on all circular dependencies, when we don't really have any plan on fixing them?
Here, `Diagnostic` is used to print operations with messages on them. So `Diagnostic` needs `ir.py` in its imports.
Then, `Operation` now needs `Diagnostic`, because it is creating a diagnostic for itself.

If we want to remove the circular dependency, we would have to merge both files, which I would be against.
So how would you see that we restructure that?

Also, are circular dependencies that bad?

Reply:
To me, circular dependencies tend to feel like you have not structured your class hierarchies properly with some classes bein more at the base (like the IR things in this project) and others being further at the edges (like the dialects here). 

But I see your point and I guess error reporting in general is pretty much impossible to do nicely with out something like this. Unless we end up moving diagnostics into ir.py, which I would not immediately reject, but we certainly need to think about it carefully.

I see your point on the TODOs that we do not address though, so let's leave as is. If we ever really want to fix this, we will find the inline imports easily.

Reply:
Personally, the reason why I don't want the diagnostics inside the `ir.py` is because the formatters and type checkers are quite bad with these files :/

Reply:
So I guess it's circular dependencies vs bad formatters xD
Let's stick to having circular dependencies for the moment and revisit this again once it becomes an isse.